Mathlib/Data/Int/Card.lean

```diff
@@ -0,0 +1,140 @@
+    dvd_iff_exists_eq_mul_left, cast_pos.2 hr, ← cast_mul, cast_lt, cast_le]
+  aesop
+
+/-- There are `⌈b / r⌉ - ⌈a / r⌉` multiples of `r` in `[a, b)`, if `a ≤ b`. -/
+theorem Ico_filter_dvd_card : ((Ico a b).filter (r ∣ ·)).card =
+    max (⌈b / (r : ℚ)⌉ - ⌈a / (r : ℚ)⌉) 0 := by
+  rw [Ico_filter_dvd_eq _ _ hr, card_map, card_Ico, toNat_eq_max]
+
+/-- There are `⌊b / r⌋ - ⌊a / r⌋` multiples of `r` in `(a, b]`, if `a ≤ b`. -/
+theorem Ioc_filter_dvd_card : ((Ioc a b).filter (r ∣ ·)).card =
+    max (⌊b / (r : ℚ)⌋ - ⌊a / (r : ℚ)⌋) 0 := by
+  rw [Ioc_filter_dvd_eq _ _ hr, card_map, card_Ioc, toNat_eq_max]
+
+lemma Ico_filter_modEq_eq (v : ℤ) : (Ico a b).filter (· ≡ v [ZMOD r]) =
+    ((Ico (a - v) (b - v)).filter (r ∣ ·)).map ⟨_, add_left_injective v⟩ := by
+  ext x
+  simp_rw [mem_map, mem_filter, mem_Ico, Function.Embedding.coeFn_mk, ← eq_sub_iff_add_eq,
+    exists_eq_right, modEq_comm, modEq_iff_dvd, sub_lt_sub_iff_right, sub_le_sub_iff_right]
+
+lemma Ioc_filter_modEq_eq (v : ℤ) : (Ioc a b).filter (· ≡ v [ZMOD r]) =
+    ((Ioc (a - v) (b - v)).filter (r ∣ ·)).map ⟨_, add_left_injective v⟩ := by
+  ext x
+  simp_rw [mem_map, mem_filter, mem_Ioc, Function.Embedding.coeFn_mk, ← eq_sub_iff_add_eq,
+    exists_eq_right, modEq_comm, modEq_iff_dvd, sub_lt_sub_iff_right, sub_le_sub_iff_right]
+
+/-- There are `⌈(b - v) / r⌉ - ⌈(a - v) / r⌉` numbers congruent to `v` mod `r` in `[a, b)`,
+if `a ≤ b`. -/
+theorem Ico_filter_modEq_card (v : ℤ) : ((Ico a b).filter (· ≡ v [ZMOD r])).card =
+    max (⌈(b - v) / (r : ℚ)⌉ - ⌈(a - v) / (r : ℚ)⌉) 0 := by
+  simp [Ico_filter_modEq_eq, Ico_filter_dvd_eq, toNat_eq_max, hr]
+
+/-- There are `⌊(b - v) / r⌋ - ⌊(a - v) / r⌋` numbers congruent to `v` mod `r` in `(a, b]`,
+if `a ≤ b`. -/
+theorem Ioc_filter_modEq_card (v : ℤ) : ((Ioc a b).filter (· ≡ v [ZMOD r])).card =
+    max (⌊(b - v) / (r : ℚ)⌋ - ⌊(a - v) / (r : ℚ)⌋) 0 := by
+  simp [Ioc_filter_modEq_eq, Ioc_filter_dvd_eq, toNat_eq_max, hr]
+
+end Int
+
+namespace Nat
+
+variable (a b : ℕ) {r : ℕ} (hr : 0 < r)
+
+lemma Ico_filter_modEq_cast {v : ℕ} : ((Ico a b).filter (· ≡ v [MOD r])).map castEmbedding =
+    (Ico ↑a ↑b).filter (· ≡ v [ZMOD r]) := by
+  ext x
+  simp only [mem_map, mem_filter, mem_Ico, castEmbedding_apply]
+  constructor
+  · simp_rw [forall_exists_index, ← coe_nat_modEq_iff]; intro y ⟨h, c⟩; subst c; exact_mod_cast h
+  · intro h; lift x to ℕ using (by linarith); exact ⟨x, by simp_all [coe_nat_modEq_iff]⟩
+
+lemma Ioc_filter_modEq_cast {v : ℕ} : ((Ioc a b).filter (· ≡ v [MOD r])).map castEmbedding =
+    (Ioc ↑a ↑b).filter (· ≡ v [ZMOD r]) := by
+  ext x
+  simp only [mem_map, mem_filter, mem_Ioc, castEmbedding_apply]
+  constructor
+  · simp_rw [forall_exists_index, ← coe_nat_modEq_iff]; intro y ⟨h, c⟩; subst c; exact_mod_cast h
+  · intro h; lift x to ℕ using (by linarith); exact ⟨x, by simp_all [coe_nat_modEq_iff]⟩
+
+/-- `Int.Ico_filter_modEq_card` restricted to natural numbers. -/
```

Comment:
This docstring is not really helpful, as the reader needs to hunt what the other lemma is saying (and tooltips won't help). Could you instead explain the content of the lemma in the docstring? 

Same thing below.

Reply:
Done.